common/djangoapps/enrollment/data.py

```diff
@@ -136,12 +136,114 @@ def update_course_enrollment(username, course_id, mode=None, is_active=None):
         return None
 
 
+def add_or_update_enrollment_attr(user_id, course_id, attributes):
+    """Set enrollment attributes for the enrollment of given user in the
+    course provided.
+
+    Args:
+        course_id (str): The Course to set enrollment attributes for.
+        user_id (str): The User to set enrollment attributes for.
+        attributes (list): Attributes to be set.
+
+    Example:
+        >>>add_or_update_enrollment_attr(
+            "Bob",
+            "course-v1-edX-DemoX-1T2015",
+            [
+                {
+                    "namespace": "credit",
+                    "name": "provider_id",
+                    "value": "hogwarts",
+                },
+            ]
+        )
+    """
+    course_key = CourseKey.from_string(course_id)
+    user = _get_user(user_id)
+    enrollment = CourseEnrollment.get_enrollment(user, course_key)
+    for attribute in attributes:
+        if not _invalid_attribute(attribute) and enrollment is not None:
+            CourseEnrollmentAttribute.set_enrollment_attribute(
+                enrollment, attribute['value'], attribute['namespace'], attribute['name']
+            )
```

Comment:
1) If the attribute dictionary is invalid, we should be returning a 400 response with an error message explaining the problem.  The caller needs to know that the request wasn't completed successfully.

2) This is going to make a database query for each attribute.  Could we use Django's [bulk_create](https://docs.djangoproject.com/en/1.4/ref/models/querysets/#django.db.models.query.QuerySet.bulk_create) to do this in one query?

3) This looks like it will add to the list of existing attributes instead of replacing that list.  I'd prefer to replace the attributes with the exact list that the caller requested.